ligo/stablecoin/spec.ligo

```diff
@@ -0,0 +1,227 @@
+// SPDX-FileCopyrightText: 2020 tqtezos
+// SPDX-License-Identifier: MIT
+
+/*
+ * Stablecoin smart-contract
+ */
+
+/* ------------------------------------------------------------- */
+
+/*
+ * Parameter
+ */
+
+type tokenId is nat
+
+
+
+type transferDescriptor is record
+  id          : tokenId
+; fromAddress : address
+; toAddress   : address
+; amount      : nat
+end
+
+type transferParams is record
+  batch    : list (transferDescriptor)
+; operator : address
+end
+
+
+
+type balanceOfRequest is record
+  id      : tokenId
+; owner   : address
+end
+
+type balanceOfResponse is record
+  request : balanceOfRequest
+; balance : nat
+end
+
+type balanceOfParams is record
+  requests : list (balanceOfRequest)
+; callback : contract (list (balanceOfResponse))
+end
+
+
+
+type getTotalSupplyResponse is record
+  id          : tokenId
+; totalSupply : nat
+end
+
+type getTotalSupplyParams is record
+  ids      : list (tokenId) // list of requests
+; callback : contract (list (getTotalSupplyResponse))
+end
+
+
+
+type getTokenMetadataResponse is record
+  id       : tokenId
+; symbol   : string
+; name     : string
+; decimals : nat
+; extras   : map (string, string)
+end
+
+type getTokenMetadataParams is record
+  id       : tokenId
+; callback : contract (list (getTokenMetadataResponse))
+end
+
+
+
+type operatorTokens is
+| AllTokens
+| SomeTokens of set (tokenId)
+
+type operatorParam is record
+  owner    : address
+; operator : address
+; tokens   : operatorTokens
+end
+
+type updateOperator is
+| AddOperator    of operatorParam
+| RemoveOperator of operatorParam
+
+type updateOperatorParams is list (updateOperator)
+
+type isOperatorResponse is record
+  operator   : operatorParam
+; isOperator : bool
+end
+
+type isOperatorParams is record
+  operator : operatorParam
+; callback : contract (isOperatorResponse)
+end
+
+/* ------------------------------------------------------------- */
+
+type selfTransferPolicy is
+| SelfTransferPermitted
+| SelfTransferDenied
+
+type operatorTransferPolicy is
+| OperatorTransferPermitted
+| OperatorTransferDenied
+
+type ownerTransferPolicy is
+| OwnerNoOp
+| OptionalOwnerHook
+| RequiredOwnerHook
+
+type customPermissionPolicy is record
+  tag       : string
+; configApi : option (address)
+end
+
+type permissionsDescriptor is record
+  self : selfTransferPolicy
+; operator : operatorTransferPolicy
+; receiver : ownerTransferPolicy
+; sender : ownerTransferPolicy
+; custom : option (customPermissionPolicy)
+end
+
+type getPermissionsDescriptorParams is
+  contract (permissionsDescriptor)
+
+/*
+ * TODO: currently LIGO does not provide capability of
+ * working with lambdas properly
+ */
+type setTransferHookParams is record
+// hook : unit -> transfer_descriptor_param contract
```

Comment:
Not sure if I am missing something, but It seems to work for me if I un-comment the `hook` field.